File: src/caffeinate.m
//
//  caffeinate.m
//  lidoff - caffeinate session management
//

#import "caffeinate.h"
#import <IOKit/pwr_mgt/IOPMLib.h>

static IOPMAssertionID assertionID = 0;
static BOOL caffeinateActive = NO;

BOOL CaffeinateStart(void) {
    if (caffeinateActive) {
        return YES;
    }
    
    CFStringRef reason = CFSTR("lidoff: lid partially closed");
    IOReturn result = IOPMAssertionCreateWithName(
        kIOPMAssertionTypePreventUserIdleDisplaySleep,
        kIOPMAssertionLevelOn,
        reason,
        &assertionID
    );
    
    if (result == kIOReturnSuccess) {
        caffeinateActive = YES;
        return YES;
    }
    
    return NO;
}

BOOL CaffeinateStop(void) {
    if (!caffeinateActive) {
        return YES;
    }
    
    IOReturn result = IOPMAssertionRelease(assertionID);
    
    if (result == kIOReturnSuccess) {
        assertionID = 0;
        caffeinateActive = NO;
        return YES;
    }
    
    return NO;
}

BOOL CaffeinateIsActive(void) {
    return caffeinateActive;
}

File: macos/caffeinate.c
#include "caffeinate.h"

#include <IOKit/pwr_mgt/IOPMLib.h>
#include <stdbool.h>

static IOPMAssertionID assertion_id = 0;
static bool caffeinate_active = false;

uint8_t CaffeinateStart(void) {
  if (caffeinate_active) {
    return 1;
  }

  IOReturn result = IOPMAssertionCreateWithName(
      kIOPMAssertionTypePreventUserIdleDisplaySleep, kIOPMAssertionLevelOn,
      CFSTR("lidoff: lid partially closed"), &assertion_id);
  if (result == kIOReturnSuccess) {
    caffeinate_active = true;
    return 1;
  }

  return 0;
}

uint8_t CaffeinateStop(void) {
  if (!caffeinate_active) {
    return 1;
  }

  IOReturn result = IOPMAssertionRelease(assertion_id);
  if (result == kIOReturnSuccess) {
    assertion_id = 0;
    caffeinate_active = false;
    return 1;
  }

  return 0;
}
